File: src/jvmTest/java/org/snakeyaml/engine/v2/scanner/FuzzYAMLRead50431Test.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.v2.scanner;

/**
 * Copyright (c) 2008, SnakeYAML
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import it.krzeminski.snakeyaml.engine.kmp.api.Load;
import it.krzeminski.snakeyaml.engine.kmp.exceptions.ScannerException;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * https://github.com/FasterXML/jackson-dataformats-text/issues/400
 * https://github.com/FasterXML/jackson-dataformats-text/pull/401
 */
@org.junit.jupiter.api.Tag("fast")
public class FuzzYAMLRead50431Test {

  private final Load load = new Load();

  @Test
  public void testIncompleteValue() {
    try {
      load.loadOne("\"\\UE30EEE");
      fail("Invalid escape code in double quoted scalar should not be accepted");
    } catch (ScannerException e) {
      assertEquals(
          "found unknown escape character E30EEE",
          e.getMessage().split("\n", 2)[0]
      );
    }
  }

  @Test
  public void testProperValue() {
    String parsed = (String) load.loadOne("\"\\U0000003B\"");
    assertEquals(1, parsed.length());
    assertEquals("\u003B", parsed);
  }

  @Test
  public void testNotQuoted() {
    String parsed = (String) load.loadOne("\\UE30EEE");
    assertEquals(8, parsed.length());
    assertEquals("\\UE30EEE", parsed);
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/scanner/FuzzYAMLRead50431Test.kt
package it.krzeminski.snakeyaml.engine.kmp.scanner

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.common.Platform
import io.kotest.common.platform
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import it.krzeminski.snakeyaml.engine.kmp.api.Load
import it.krzeminski.snakeyaml.engine.kmp.exceptions.ScannerException

class FuzzYAMLRead50431Test : FunSpec({

    val load = Load()

    test("incomplete value should throw ScannerException") {
        shouldThrow<ScannerException> {
            load.loadOne("\"\\UE30EEE")
        }.also {
            // There's a difference in the error message between JVM and other platforms.
            // TODO: understand and fix in https://github.com/krzema12/snakeyaml-engine-kmp/issues/550
            it.message shouldContain if (platform == Platform.JVM) {
                "found unknown escape character E30EEE"
            } else {
                "found unexpected end of stream"
            }
        }
    }

    test("proper value should be parsed correctly") {
        val parsed = load.loadOne("\"\\U0000003B\"") as String
        parsed.length shouldBe 1
        parsed shouldBe "\u003B"
    }

    test("not quoted value should be parsed as literal string") {
        val parsed = load.loadOne("\\UE30EEE") as String
        parsed.length shouldBe 8
        parsed shouldBe "\\UE30EEE"
    }
})
